### ISSUE TITLE
Compile Parse methods into IL

### ISSUE THREAD

**Issue by KvanTTT** [AvaloniaUI/Avalonia#5392]
It reproduces at `CanvasPage` in ControlCatalog sample app. I suggest the following optimizations:

`RelativePoint.Parse("100%,100%");` -> `new RelativePoint(1.0, 1.0, RelativeUnit.Absolute)`
`Point.Parse("0,0")` -> `new Point(0, 0)`
`Color.Parse("Black")` -> `new Color(1, 0, 0, 0)`

And other objects declared in `Avalonia.Media`.

It is also possible to get rid of some excess invocations if use caching of immutable objects.

**Comment by jp2masa** [comment]
#5053
#5065
#5075
#5088

From your list I think `RelativePoint` is the only left.

**Comment by jp2masa** [comment]
Or are you referring to code-behind calls?

**Comment by KvanTTT** [comment]
Yes, I mean code-behind calls. On the latest master I have the following code for CanvasPage (disassembled by dnSpy):

```cs
LinearGradientBrush linearGradientBrush = new LinearGradientBrush();
linearGradientBrush.StartPoint = RelativePoint.Parse("0%,0%");
linearGradientBrush.EndPoint = RelativePoint.Parse("100%,100%");
GradientStop gradientStop = new GradientStop
{
	Offset = 0.0,
	Color = Color.Parse("Black")
};
```

I think such optimization may decrease the time of app launching especially with a big number of Canvas elements.

**Comment by jp2masa** [comment]
For now I would recommend declaring those as XAML resources, it's probably faster because it will be compiled.

The problem here is that what you're asking for would require IL weaving I think.

Possible alternatives would be for example:

- A roslyn "code fix" which would replace that with the compiled data, but it would be less readable, and you would lose the original string
- A roslyn source generator, which would parse the text specified on an attribute and store it somewhere

Both alternatives aren't portable between languages, so not sure if it's worth it.

**Comment by kekekeks** [comment]
@jp2masa [CanvasPage.xaml.cs](https://github.com/AvaloniaUI/Avalonia/blob/1a39210febf92e9e87b3ebacca9b56d99e660913/samples/ControlCatalog/Pages/CanvasPage.xaml.cs) is almost empty, so those values come from XAML.

@KvanTTT you are probably using an outdated version of Avalonia.
The output of `ilspycmd ControlCatalog.dll |egrep -o '[A-Za-z]+\.Parse' |sort -n |uniq` for 0.10 is
```
char.Parse
Classes.Parse
Geometry.Parse
IterationCount.Parse
KeyGesture.Parse
PixelRect.Parse
RelativePoint.Parse
SolidColorBrush.Parse
StreamGeometry.Parse
TextDecorationCollection.Parse
```

Some of those could indeed be parsed ad compile time (probably except for `StreamGeometry`). Feel free to extend [ AvaloniaXamlIlLanguageParseIntrinsics.TryConvert](https://github.com/AvaloniaUI/Avalonia/blob/15968cb2c0aa115545e6766ee321ffd9eaa6d8d0/src/Markup/Avalonia.Markup.Xaml.Loader/CompilerExtensions/AvaloniaXamlIlLanguageParseIntrinsics.cs#L17) if any of those are degrading the perf of your app

**Comment by KvanTTT** [comment]
@kekekeks you are right, it was an outdated version, sorry for the confusion. But some methods still are not optimized indeed as you noticed. Why `Geometry`, `StreamGeometry` and probably `Classes` can not be optimized in the compiler time? Also, what do you think about brushes `(IBrush)new BrushConverter().ConvertFrom(context, CultureInfo.InvariantCulture, "Blue");`?

**Comment by Gillibald** [comment]
Geometry can be compiled via https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Visuals/Platform/IGeometryContext.cs

**Comment by KvanTTT** [comment]
I want to implement some compilations.

**Comment by kekekeks** [comment]
@Gillibald we probably also want to generate cached instances for immutable and expensive to create type.

**Comment by MarchingCube** [comment]
@KvanTTT I think I might have a branch somewhere with `Classes` parsed at compile time somewhere. I didn't implement parsing of other types since I was using profiling data from my app to prioritize types that mattered for my app. But I think there is nothing blocking us from writing the remainder of parsers for other numeric types.

Geometry is slightly more complicated since we need to have a cache like mentioned above. Plus we need to make sure that such types are actually immutable.
